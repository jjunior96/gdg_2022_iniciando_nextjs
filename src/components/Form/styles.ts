import styled, { css } from 'styled-components';

import * as ButtonComponent from 'components/Button/styles';
import * as InputComponent from 'components/Input/styles';

export const Container = styled.form``;

export const Content = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    flex: 1;

    width: 100%;
    max-width: ${theme.grid.container};
    min-width: ${theme.grid.container};

    ${InputComponent.Container} {
      margin-bottom: ${theme.spacings.xxsmall};
    }

    ${ButtonComponent.Container} {
      margin-top: ${theme.spacings.xxsmall};
    }

    label {
      margin-top: ${theme.spacings.xsmall};
    }

    button[type='submit'] {
      margin: ${theme.spacings.small} 0;
    }

    @media (max-width: 960px) {
      min-width: 100%;
    }

    @media (max-width: 500px) {
      max-width: 100%;
      min-width: 100%;
    }
  `}
`;
