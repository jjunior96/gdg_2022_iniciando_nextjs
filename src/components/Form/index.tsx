import { FormHTMLAttributes } from 'react';

import * as S from './styles';

export type FormProps = {
  children: React.ReactNode;
} & FormHTMLAttributes<HTMLFormElement>;

const Form = ({ children, onSubmit }: FormProps) => {
  return (
    <S.Container onSubmit={onSubmit}>
      <S.Content>{children}</S.Content>
    </S.Container>
  );
};

export default Form;
