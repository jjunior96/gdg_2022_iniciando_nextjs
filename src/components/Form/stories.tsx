import { action } from '@storybook/addon-actions';
import { Story, Meta } from '@storybook/react/types-6-0';

import Button from 'components/Button';
import Heading from 'components/Heading';
import Input from 'components/Input';

import Form, { FormProps } from '.';

export default {
  title: 'Form/Form',
  component: Form
} as Meta;

export const Default: Story<FormProps> = (args) => (
  <div
    style={{
      width: '40rem',
      height: '25rem',
      margin: '0 auto'
    }}
  >
    <Form
      {...args}
      onSubmit={(e) => {
        e.preventDefault();
        action('form submitted')(e);
      }}
    >
      <Heading>User info</Heading>
      <Input name="name" placeholder="Insert user name" label="Name" />{' '}
      <Input
        name="email"
        type="email"
        placeholder="Insert user email"
        label="E-mail"
      />
      <Button type="submit">Enter</Button>
    </Form>
  </div>
);
