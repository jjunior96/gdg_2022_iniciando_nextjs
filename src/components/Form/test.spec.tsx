import { fireEvent, screen, waitFor } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import Form from '.';

const onSubmit = jest.fn((e) => e.preventDefault());

describe('<Form />', () => {
  it('should render the Form', () => {
    const { container } = renderWithTheme(
      <Form aria-label="User info" name="userinfo">
        <input aria-label="username" placeholder="Username" />
      </Form>
    );

    expect(
      screen.getByRole('textbox', { name: /username/i }).parentElement
    ).toHaveStyle({
      display: 'flex',
      'flex-direction': 'column',

      width: '100%'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should submit when onSubmit method is called', async () => {
    renderWithTheme(
      <Form aria-label="User info" name="userinfo" onSubmit={onSubmit}>
        <input aria-label="username" placeholder="Username" />

        <button type="submit">Salve</button>
      </Form>
    );

    const buttonSubmit = screen.getByRole('button', { name: /salve/i });

    fireEvent.click(buttonSubmit);

    await waitFor(async () => {
      expect(onSubmit).toHaveBeenCalled();
    });
  });
});
