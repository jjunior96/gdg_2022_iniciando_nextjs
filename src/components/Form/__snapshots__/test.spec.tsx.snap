// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render the Form 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  max-width: 96rem;
  min-width: 96rem;
}

.c0 label {
  margin-top: 1.6rem;
}

.c0 button[type='submit'] {
  margin: 2.4rem 0;
}

@media (max-width:960px) {
  .c0 {
    min-width: 100%;
  }
}

@media (max-width:500px) {
  .c0 {
    max-width: 100%;
    min-width: 100%;
  }
}

<form
  class=""
>
  <div
    class="c0"
  >
    <input
      aria-label="username"
      placeholder="Username"
    />
  </div>
</form>
`;
